components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/SimpleObjectEncoder.scala

```diff
@@ -1,47 +1,101 @@
-    case (`doubleClass`, doubleValue: Double) =>
-      fromDouble(doubleValue).map(_.validNel).getOrElse(s"Could not encode $doubleValue as json.".invalidNel)
-    case (`booleanClass`, booleanValue: Boolean) =>
-      fromBoolean(booleanValue).validNel
-    case (`stringClass`, stringValue: String) =>
-      fromString(stringValue).validNel
-    case (klass, value) if value.getClass == klass.klass =>
-      s"No encoding logic for $typ.".invalidNel
-    case (klass, value) =>
-      s"Mismatched class and value: $klass and $value".invalidNel
+  private val intClass        = Typed.typedClass[Int]
+  private val shortClass      = Typed.typedClass[Short]
+  private val longClass       = Typed.typedClass[Long]
+  private val floatClass      = Typed.typedClass[Float]
+  private val doubleClass     = Typed.typedClass[Double]
+  private val booleanClass    = Typed.typedClass[Boolean]
+  private val stringClass     = Typed.typedClass[String]
+  private val byteClass       = Typed.typedClass[Byte]
+  private val bigIntegerClass = Typed.typedClass[BigInteger]
+  private val bigDecimalClass = Typed.typedClass[java.math.BigDecimal]
+
+  def encodeValue(value: Any): ValidatedNel[String, Json] = value match {
+    case null                        => Json.Null.validNel
+    case value: Int                  => fromInt(value).validNel
+    case value: Short                => fromInt(value).validNel
+    case value: Long                 => fromLong(value).validNel
+    case value: Boolean              => fromBoolean(value).validNel
+    case value: String               => fromString(value).validNel
+    case value: Byte                 => fromInt(value).validNel
+    case value: BigInteger           => fromBigInt(value).validNel
+    case value: java.math.BigDecimal => fromBigDecimal(value).validNel
+    case value: Float =>
+      fromFloat(value).map(_.validNel).getOrElse(s"Could not encode $value as json.".invalidNel)
+    case value: Double =>
+      fromDouble(value).map(_.validNel).getOrElse(s"Could not encode $value as json.".invalidNel)
+
+    case vals: java.util.Collection[_] =>
+      val encodedValues = vals.asScala.map(elem => encodeValue(elem)).toList.sequence
+      encodedValues.map(values => Json.fromValues(values))
+    case vals: java.util.Map[_, _] =>
+      val encodedMap: Validated[NonEmptyList[String], List[(String, Json)]] = vals.asScala
+        .map { case (key, value) =>
+          encodeValue(key).andThen(encodedKey =>
+            encodedKey.asString match {
+              case Some(encodedKeyString) => encodeValue(value).map(encodedValue => encodedKeyString -> encodedValue)
+              case None                   => s"Failed to encode Record key '$encodedKey' as String".invalidNel
+            }
+          )
+        }
+        .toList
+        .sequence
+
+      encodedMap.map(values => Json.fromFields(values))
+    case value =>
+      s"No encoding logic for value $value of class ${value.getClass}".invalidNel
```

Comment:
I would write `Encoding of value [$value] of class [${value.getClass.getName}] is not supported`